### DISCUSSION TITLE
Why do I get "fatal: not a git repository: '.'" when using git python in a pre-receive git hook?

### DISCUSSION THREAD

**Discussion by pwhipp** [gitpython-developers/GitPython#1207] [Q&A]
I'm not sure if this is an issue with GitPython or not but I'm reporting it here in case it is and in the hope you can shed some light on the problem.

I am developing an application that uses a pre-receive hook. The hook needs to generate a map of the added/modified/deleted files relating to the submitted change and, as I've used GitPython before it seemed natural to use it here. The pre-receive bash script dispatches to my code thus:
```
#!/usr/bin/env bash
source /home/paul/wk/sphinx_man/venv/bin/activate
exec python -m protrep.pre_receive
```
And the pre_receive looks something like this (stripped down to show the issue):
```
import os
import sys
import logging
from collections import defaultdict

from git import Repo

from protrep.settings import LOGGING

logging.basicConfig(**LOGGING)


def pre_receive():
    update_map = read_update_map(sys.stdin)

    main_update = update_map.get('refs/heads/main', None)
    if main_update is None:
        logging.debug("pre-receive: No main reference update. Skipping processing for now")
        return

    repo = Repo(os.path.dirname(os.getcwd()))

    repo.git.status()


def read_update_map(f):
    return {ref: {'old': old, 'new': new}
            for old, new, ref in [line.split() for line in f]}


if __name__ == "__main__":
    pre_receive()
```
The error I get is:
```
/tmp/test/repo $ git push
Enumerating objects: 4, done.
Counting objects: 100% (4/4), done.
Delta compression using up to 12 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 274 bytes | 274.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0)
remote: Traceback (most recent call last):
remote:   File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
remote:     return _run_code(code, main_globals, None,
remote:   File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
remote:     exec(code, run_globals)
remote:   File "/home/paul/wk/sphinx_man/sphinx_man/protrep/pre_receive.py", line 53, in <module>
remote:     pre_receive()
remote:   File "/home/paul/wk/sphinx_man/sphinx_man/protrep/pre_receive.py", line 29, in pre_receive
remote:     repo.git.status()
remote:   File "/home/paul/wk/sphinx_man/venv/lib/python3.9/site-packages/git/cmd.py", line 545, in <lambda>
remote:     return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
remote:   File "/home/paul/wk/sphinx_man/venv/lib/python3.9/site-packages/git/cmd.py", line 1011, in _call_process
remote:     return self.execute(call, **exec_kwargs)
remote:   File "/home/paul/wk/sphinx_man/venv/lib/python3.9/site-packages/git/cmd.py", line 828, in execute
remote:     raise GitCommandError(command, status, stderr_value, stdout_value)
remote: git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
remote:   cmdline: git status
remote:   stderr: 'fatal: not a git repository: '.''
To /tmp/test/POT
 ! [remote rejected] main -> main (pre-receive hook declined)
error: failed to push some refs to '/tmp/test/POT'
```
When I invoke the same command in a shell on the same repository it works fine. Obviously I'll probably be using diff-tree etc. in the real application but this same error occurs whatever command I invoke, in spite of the fact that I can copy paste the command into the shell and it works:
```
(sphinx_man) /tmp/test/POT $ ipython
Python 3.9.2 (default, Feb 20 2021, 20:56:08) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from git import Repo

In [2]: import os

In [3]: repo = Repo(os.path.dirname(os.getcwd()))
---------------------------------------------------------------------------
InvalidGitRepositoryError                 Traceback (most recent call last)
<ipython-input-3-1722a69e1ca0> in <module>
----> 1 repo = Repo(os.path.dirname(os.getcwd()))

~/wk/sphinx_man/venv/lib/python3.9/site-packages/git/repo/base.py in __init__(self, path, odbt, search_parent_directories, expand_vars)
    179 
    180         if self.git_dir is None:
--> 181             raise InvalidGitRepositoryError(epath)
    182 
    183         self._bare = False

InvalidGitRepositoryError: /tmp/test

In [4]: repo = Repo(os.getcwd())

In [5]: repo.git.status()
Out[5]: 'On branch main\nnothing to commit, working tree clean'
```
Note that the shell was not running in the .git folder hence the slight tweak (left in to show that it would generate a different error anyway).

The logging debug log shows:
```
2021/03/22 14:30:02 :DEBUG: Popen(['git', 'status'], cwd=/tmp/test/POT, universal_newlines=False, shell=None, istream=None)
```
which presumably comes from GitPython and looks like it should end up doing the right thing.

Any idea what is going on?

**Comment by Byron** [comment]
Thanks for the detailed logs!
Did you try to create a repository like `repo = Repo()`? This automatically uses the current working directory and no further path adjustments should be necessary.

**Comment by pwhipp** [comment]
I have now and I get the same result.

I'm thinking this may have something to do with the way that the pre-receive environment is isolated and set up to contain the new objects (which are not kept if it fails) but I can't find any documentation on that.

**Comment by Byron** [comment]
I think it would be good to know the gitpython version in use here.

Furthermore you could add this to the bash script:

```
env | grep -i git
pwd
ls -l # or specifically the git repository 
```

I have the feeling the git directory is a file.

**Reply by pwhipp** [reply]
GitPython==3.1.13

The additions to the bash script won't help because of the process invocation but I added: ``logging.debug(f"pre-receive: working in {os.path.dirname(os.getcwd())}")`` to the script and it reports: 2021/03/23 05:38:13 :DEBUG: pre-receive: working in /tmp/test/POT.

In https://git-scm.com/docs/git-receive-pack, there is a note about the 'quarantine environment' which may be relevant.

I'm thinking I might ultimately need to use my own git command (via send-pack/receive-pack(exec)) because I need to serialize and offload the potentially time consuming validations associated with the application. That said, I still need the pre-receive hook to work.

**Reply by Byron** [reply]
Even though I don't think anything was fixed since then, 3.1.14 might be worth trying out.

That said, without further information I don't think there is anything I can do. Note that environment variables are set differently in git hooks and they might affect GitPython. Additionally not all git repositories are made equal which could be another factor.
